### PR TITLE
Add non-supervised mode to ExpertMixture

### PR DESCRIPTION
**Pull request by jschueller**
and enrich doc, test

### CODE REVIEW

File: python/src/ExpertMixture_doc.i.in [hunk 1]
```diff
@@ -1,45 +1,68 @@
...
+
+where the :math:`N` classes are defined by the classifier.
+
+In supervised mode the classifier partitions the input and output space at once
+
+.. math::
+
+    \vect{z} = (\vect{x}, f(\vect{x}))
+
+else in non-supervised mode only the input space is partitioned
```
Comment (dbarbier) [inline python/src/ExpertMixture_doc.i.in]: s/else/whereas/

File: python/src/ExpertMixture_doc.i.in [hunk 2]
```diff
@@ -1,45 +1,68 @@
 %feature("docstring") OT::ExpertMixture
 "Expert mixture defining a piece wise function according to a classifier.
 
-Available constructors:
-    ExpertMixture(*basis, classifier*)
+This implements an expert mixture which is a piece wise function :math:`f`
```
Comment (dbarbier) [inline python/src/ExpertMixture_doc.i.in]: s/piece wise/piecewise/

File: python/src/ExpertMixture_doc.i.in [hunk 3]
```diff
@@ -1,45 +1,68 @@
...
 
 Parameters
 ----------
 basis : :class:`~openturns.Basis`
     A basis which is a collection of :class:`~openturns.Function`.
 classifier : :class:`~openturns.Classifier`
     A classifier
+supervised : bool (default=True)
+    In supervised mode, the classifer partitions the space
+    of :math:`(\vect(x), f(\vect(x)))`, else it has the dimension of *x*.
```
Comment (dbarbier) [inline python/src/ExpertMixture_doc.i.in]: I do not understand the end of the sentence.

File: python/src/ExpertMixture_doc.i.in [hunk 4]
```diff
@@ -1,45 +1,68 @@
 %feature("docstring") OT::ExpertMixture
 "Expert mixture defining a piece wise function according to a classifier.
 
-Available constructors:
-    ExpertMixture(*basis, classifier*)
+This implements an expert mixture which is a piece wise function :math:`f` 
+defined by the collection of functions :math:`(f_i)_{i=1, \ldots, N}` given in
+*basis* and according to the classifier *classifier*:
```
Comment (dbarbier) [inline python/src/ExpertMixture_doc.i.in]: You do not repeat basis, you could do the same for classifier.

File: python/src/ExpertMixture_doc.i.in [hunk 5]
```diff
@@ -1,45 +1,68 @@
...
+    \vect{z} = \vect{x}
 
 Parameters
 ----------
 basis : :class:`~openturns.Basis`
     A basis which is a collection of :class:`~openturns.Function`.
 classifier : :class:`~openturns.Classifier`
     A classifier
+supervised : bool (default=True)
+    In supervised mode, the classifer partitions the space
```
Comment (dbarbier) [inline python/src/ExpertMixture_doc.i.in]: s/classifer/classifier/

File: lib/src/Base/Func/ExpertMixture.cxx [hunk 6]
```diff
@@ -34,22 +34,34 @@ ExpertMixture::ExpertMixture()
   : EvaluationImplementation()
   , experts_()
   , classifier_()
+  , supervised_(false)
```
Comment (dbarbier) [inline lib/src/Base/Func/ExpertMixture.cxx]: In the other constructor, supervised_ is true by default, it looks awkward to have false here.

File: python/src/ExpertMixture_doc.i.in (hunk 2 above)
Comment (dbarbier) [inline python/src/ExpertMixture_doc.i.in]: Please add a trailing dot.

File: lib/src/Base/Func/ExpertMixture.cxx [hunk 7]
```diff
@@ -114,17 +126,19 @@ Point ExpertMixture::operator() (const Point & inP) const
   const UnsignedInteger outputDimension = getOutputDimension();
   const UnsignedInteger size = experts_.getSize();
   UnsignedInteger bestClass = 0;
-  // Build the point (x, f(x)) for the first class and grade it according to the classifier
+  // Build the point z for the first class and grade it according to the classifier
   Point mixedPoint(inP);
   Point bestValue(experts_[0](inP));
```
Comment (regislebrun) [inline lib/src/Base/Func/ExpertMixture.cxx]: const Point bestValue(experts_[0](inP));

File: lib/src/Base/Func/ExpertMixture.cxx [hunk 8]
```diff
@@ -114,17 +126,19 @@ Point ExpertMixture::operator() (const Point & inP) const
...
   Scalar bestGrade = classifier_.grade(mixedPoint, bestClass);
   LOGDEBUG(OSS() << "Class index=" << 0 << ", grade=" << bestGrade << ", value=" << bestValue);
   for (UnsignedInteger classIndex = 1; classIndex < size; ++classIndex)
   {
-    // Build the point (x, f(x)) for each other class and grade it according to the classifier
+    // Build the point z for each other class and grade it according to the classifier
     const Point localValue(experts_[classIndex](inP));
-    for (UnsignedInteger i = 0; i < outputDimension; ++i) mixedPoint[inputDimension + i] = localValue[i];
+    if (supervised_)// z=(x, f(x))
+      for (UnsignedInteger i = 0; i < outputDimension; ++i) mixedPoint[inputDimension + i] = localValue[i];
```
Comment (regislebrun) [inline lib/src/Base/Func/ExpertMixture.cxx]: std::copy(localValue.begin(), localValue.end(), mixedPoint.begin()+inputDimension);

File: lib/src/Base/Func/ExpertMixture.cxx (hunk 7 above)
Comment (jschueller) [inline lib/src/Base/Func/ExpertMixture.cxx]: cant do

File: lib/src/Base/Func/ExpertMixture.cxx (hunk 8 above)
Comment (jschueller) [inline lib/src/Base/Func/ExpertMixture.cxx]: ok

File: lib/src/Base/Func/openturns/ExpertMixture.hxx [hunk 9]
```diff
@@ -47,7 +47,8 @@ public:
 
   /** Constructor */
   ExpertMixture(const Basis & experts,
-                const Classifier & classifier);
+                const Classifier & classifier,
+                const Bool supervised = true);
```
Comment (jschueller) [inline lib/src/Base/Func/openturns/ExpertMixture.hxx]: It's really tied to the classifier here, so changing it in the RM is not enough, and will throw.

File: python/test/t_ExpertMixture_std.py [hunk 10]
```diff
@@ -1,38 +1,41 @@
...
-# Create local experts
-experts = Basis()
-experts.add(SymbolicFunction("x", "-x"))
-experts.add(SymbolicFunction("x", "x"))
 
-# Create a mixture of experts
-moe = ExpertMixture(experts, classifier)
-moeNMF = Function(moe)
+# non-supervised mode (2d)
+f1 = ot.SymbolicFunction(['x1', 'x2'], ['-8'])
```
Comment (jschueller) [inline python/test/t_ExpertMixture_std.py]: In what way ? I test both modes.

File: lib/src/Base/Func/ExpertMixture.cxx [hunk 11]
```diff
@@ -114,17 +126,19 @@ Point ExpertMixture::operator() (const Point & inP) const
...
   Scalar bestGrade = classifier_.grade(mixedPoint, bestClass);
   LOGDEBUG(OSS() << "Class index=" << 0 << ", grade=" << bestGrade << ", value=" << bestValue);
   for (UnsignedInteger classIndex = 1; classIndex < size; ++classIndex)
   {
-    // Build the point (x, f(x)) for each other class and grade it according to the classifier
+    // Build the point z for each other class and grade it according to the classifier
     const Point localValue(experts_[classIndex](inP));
-    for (UnsignedInteger i = 0; i < outputDimension; ++i) mixedPoint[inputDimension + i] = localValue[i];
+    if (supervised_)// z=(x, f(x))
+      std::copy(localValue.begin(), localValue.end(), mixedPoint.begin() + inputDimension);
```
Comment (jschueller) [inline lib/src/Base/Func/ExpertMixture.cxx]: No, you still have to evaluate it in both modes to return the result.

File: lib/src/Base/Func/ExpertMixture.cxx [hunk 12]
```diff
@@ -150,12 +164,13 @@ Sample ExpertMixture::operator() (const Sample & inS) const
...
   {
-    // Build the point (x, f(x)) for each other class and grade it according to the classifier
+    // Build the point z for each other class and grade it according to the classifier
     Sample mixedSample(inS);
     // Here is the evaluation of the expert over a sample, benefiting from possible
     // parallelism/vectorization
     const Sample localValues(experts_[classIndex](inS));
-    mixedSample.stack(localValues);
+    if (supervised_)// z=(x, f(x))
+      mixedSample.stack(localValues);
```
Comment (jschueller) [inline lib/src/Base/Func/ExpertMixture.cxx]: Idem

File: python/src/ExpertMixture_doc.i.in [hunk 13]
```diff
@@ -10,6 +10,9 @@ basis : :class:`~openturns.Basis`
     A basis which is a collection of :class:`~openturns.Function`.
 classifier : :class:`~openturns.Classifier`
     A classifier
+supervised : bool, default=True
+    In supervised mode, the classifer partitions the database
+    of :math:`(x_i, f(x_i))`, else it has the dimension of *x*.
```
Comment (regislebrun) [inline python/src/ExpertMixture_doc.i.in]: else it partitions the database of :math:`(x_i)` only.

File: lib/src/Base/Func/ExpertMixture.cxx (hunk 11 above)
Comment (sofianehaddad) [inline lib/src/Base/Func/ExpertMixture.cxx]: No you have to return the bestValue but could compute it once the classification is done. This last one does not use the *localValue*

File: lib/src/Base/Func/ExpertMixture.cxx [hunk 14]
```diff
@@ -114,17 +126,19 @@ Point ExpertMixture::operator() (const Point & inP) const
...
-  mixedPoint.add(bestValue);
+  if (supervised_)// z=(x, f(x))
+    mixedPoint.add(bestValue);
   Scalar bestGrade = classifier_.grade(mixedPoint, bestClass);
   LOGDEBUG(OSS() << "Class index=" << 0 << ", grade=" << bestGrade << ", value=" << bestValue);
   for (UnsignedInteger classIndex = 1; classIndex < size; ++classIndex)
   {
-    // Build the point (x, f(x)) for each other class and grade it according to the classifier
+    // Build the point z for each other class and grade it according to the classifier
     const Point localValue(experts_[classIndex](inP));
```
Comment (jschueller) [inline lib/src/Base/Func/ExpertMixture.cxx]: No, it's returned too via bestValue when supervised_ is false.

File: python/src/ExpertMixture_doc.i.in [hunk 15]
```diff
@@ -1,45 +1,69 @@
...
+>>> print(moe([0.1]))
+[0.1]
 
 Notes
 -----
-This implements an expert mixture which is a piece wise function :math:`f` 
-defined by the collection of functions :math:`(f_i)_{i=1, \ldots, N}` given in
-*basis* and according to the classifier *classifier*:
+The number of experts may be less or equal to the number of classes of the
+classifier.
```
Comment (jschueller) [inline python/src/ExpertMixture_doc.i.in]: It should, but the code allows to provide less experts than the number of classes actually returned by the classifier (and there's no way to know it in advance).

Comment (regislebrun) [inline python/src/ExpertMixture_doc.i.in]: @jschueller @sofianehaddad Yes, the number of experts must match the number of classes. If the same expert can be used for many classes, the correct way to do this is to merge these classes into a unique class.

File: lib/src/Base/Func/ExpertMixture.cxx (hunk 14 above)
Comment (sofianehaddad) [inline lib/src/Base/Func/ExpertMixture.cxx]: I was thinking about something like https://gist.github.com/sofianehaddad/359f4e96329d7700384657f2212bff18

File: python/test/t_ExpertMixture_std.py (hunk 10 above)
Comment (sofianehaddad) [inline python/test/t_ExpertMixture_std.py]: As we are introducing the non-supervised mode it should *correctly* be tested :
 - at least 2 different use case
 - on a sample and not only two points

File: lib/src/Base/Func/openturns/ExpertMixture.hxx (hunk 9 above)
Comment (sofianehaddad) [inline lib/src/Base/Func/openturns/ExpertMixture.hxx]: yes I mean only at initialization
At this level we had various incoherence in the library (sometimes we get RM keys, sometimes no)

File: python/test/t_ExpertMixture_std.py (hunk 10 above)
Comment (jschueller) [inline python/test/t_ExpertMixture_std.py]: I added test on samples.

File: lib/src/Base/Func/openturns/ExpertMixture.hxx (hunk 9 above)
Comment (jschueller) [inline lib/src/Base/Func/openturns/ExpertMixture.hxx]: Here I think we cannot use a key.

File: lib/src/Base/Func/ExpertMixture.cxx (hunk 11 above)
Comment (jschueller) [inline lib/src/Base/Func/ExpertMixture.cxx]: Oh, you're right. Though I think it's best to leave as is as this complicates the code.

File: python/src/ExpertMixture_doc.i.in (hunk 15 above)
Comment (regislebrun) [inline python/src/ExpertMixture_doc.i.in]: @jschueller So something is missing in the Classifier class. The number of experts *must* be equal to the number of classes, and a Classifier *should* know its number of classes! BTW, there can be more experts than the number of classes if this number is not known and not checked.

Comment (jschueller) [inline python/src/ExpertMixture_doc.i.in]: Ok, added.

Comment (regislebrun) [inline python/src/ExpertMixture_doc.i.in]: +1

File: lib/src/Base/Func/ExpertMixture.cxx (hunk 11 above)
Comment (sofianehaddad) [inline lib/src/Base/Func/ExpertMixture.cxx]: Not sure this complicates the code as we separate into two specific supervised/non supervised operators. Btw we can merge it as is it 
I will provide later a new PR implementing https://gist.github.com/sofianehaddad/359f4e96329d7700384657f2212bff18

Comment (jschueller) [inline lib/src/Base/Func/ExpertMixture.cxx]: ok you win: 1791875

**Comment by dbarbier** [comment]
evaluateSupervised(Sample) and evaluateSupervised(Point) have spurious differences in loops over classIndex, this is confusing (ditto for evaluateNonSupervised).  I prefer the Sample version, starting from 0.

File: lib/src/Base/Func/ExpertMixture.cxx [hunk 16]
```diff
@@ -139,25 +158,60 @@ Point ExpertMixture::operator() (const Point & inP) const
...
 }
 
+Point ExpertMixture::evaluateNonSupervised(const Point & inP) const
+{
+  const UnsignedInteger inputDimension = getInputDimension();
+  if (inP.getDimension() != inputDimension) throw InvalidArgumentException(HERE) << "Error: expected a point of dimension=" << inputDimension << " and got a point of dimension=" << inP.getDimension();
+  const UnsignedInteger size = experts_.getSize();
+  UnsignedInteger bestClass = 0;
+  // Build the point z for the first class and grade it according to the classifier
+  Point mixedPoint(inP);
```
Comment (dbarbier) [inline lib/src/Base/Func/ExpertMixture.cxx]: mixedPoint is useless here, use inP instead.

Comment (jschueller) [inline lib/src/Base/Func/ExpertMixture.cxx]: ok

**Comment by dbarbier** [comment]
Ok, thanks.

**Pull request merged**